### PR TITLE
[sparsity][refactor] Changing linear row/col control

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60728 [quant][sparsity] Generic convert function
* #59771 [sparsity] Lambda Scheduler
* #59770 [sparsity] Base sparsity level scheduler class
* #58955 [sparsity] WeightNormSparsifier
* #58707 [sparsity][refactor] Import factoring out
* #58704 [sparsity] Sparsifier class
* #58705 [sparsity] Sparsity parametrization
* **#60850 [sparsity][refactor] Changing linear row/col control**
* #60887 [sparsity] Add sparsity tests to run_test.py

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465900](https://our.internmc.facebook.com/intern/diff/D29465900)

### CODE REVIEW

File: torch/ao/nn/sparse/quantized/utils.py [hunk 1]
```diff
@@ -19,20 +19,20 @@ class QNNPACKLinearBlockSparsePattern:
 
     def __init__(self, row_block_size=1, col_block_size=4):
         assert(is_valid_linear_block_sparse_pattern(row_block_size, col_block_size))
-        QNNPACKLinearBlockSparsePattern.rlock.acquire()
-        QNNPACKLinearBlockSparsePattern.prev_row_block_size = QNNPACKLinearBlockSparsePattern.row_block_size
-        QNNPACKLinearBlockSparsePattern.prev_col_block_size = QNNPACKLinearBlockSparsePattern.col_block_size
-        QNNPACKLinearBlockSparsePattern.row_block_size = row_block_size
-        QNNPACKLinearBlockSparsePattern.col_block_size = col_block_size
+        LinearBlockSparsePattern.rlock.acquire()
```
Comment (raghuramank100) [inline torch/ao/nn/sparse/quantized/utils.py:22]: Why is there an rlock acquire here? From point of view of modularity, should this be in C++ as part of the kernels? Do we need this for both FBGEMM and QNNPACK?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments

File: torch/ao/nn/sparse/quantized/utils.py (hunk 1 above)
Comment (z-a-f) [reply torch/ao/nn/sparse/quantized/utils.py:22]: This is a lock for thread safety on python level.

Comment (raghuramank100) [reply torch/ao/nn/sparse/quantized/utils.py:22]: cc @vkuzo  Wondering if there are any other side effects to this?

Comment (vkuzo) [reply torch/ao/nn/sparse/quantized/utils.py:22]: This PR is just renaming things, the code was added by @kimishpatel and is not really related to this PR afaik.  @kimishpatel , would you mind chiming in with some more context on why this is here?

Agreed that it's unexpected to see this in a sparsity utils file, ideally this stuff would be in a lower level multithreading utility somewhere.  If it is needed here, can we add a comment describing why?

Comment (z-a-f) [reply torch/ao/nn/sparse/quantized/utils.py:22]: I feel like I was explaining this for some time already. Singletons in python are not thread-safe. To make them thread safe, it's a common practice to add locks in there. Adding explanation on that is fine, which I could do in a separate PR. But I am still confused about why it's blocking landing of the current refactoring PR.

Comment (vkuzo) [reply torch/ao/nn/sparse/quantized/utils.py:22]: > But I am still confused about why it's blocking landing of the current refactoring PR.

+1, I was only responding because Raghu tagged me, but this discussion is not related to the goal of this PR.

Comment (kimishpatel) [reply torch/ao/nn/sparse/quantized/utils.py:22]: @vkuzo @z-a-f @raghuramank100, originally I added this as an C++ equivalent RAII object for context management. Context being managed here is the block pattern of sparsity. Reason we need context management: At the time of this code writing, we did not have sparsity workflow (as mentioned in the comment) and we were piggy-backing on quantization workflow. Since quantization configs dont have a way to specify sparsity pattern I added this context manager that said "now sparsity pattern is *this* and for all SparseDynamicLinearPackedParam that you create, use the specified sparsity captured by the context". This was need because QNNPACK supported 1x4 and 8x1 unliked fbgemm that supported only 1x4. So we needed a way to specify which sparsity pattern do we want during convert phase of the sparsity workflow. Let me know if this adds the needed context or made it worse?

Once we have sparsity workflow whose config captures sparsity pattern we should not need this.

Comment (z-a-f) [reply torch/ao/nn/sparse/quantized/utils.py:22]: @raghuramank100 Should we add comments in file to avoid confusion? As we discussed in person this is a temporary measure until we are done (the same as @kimishpatel mentioned), so I think we don't need to explain it. However, it might be useful for posterity and paper-trail.

Either way, if there are no other concerns, I would appreciate a stamp :) If comments needed, I will send it as a separate PR

Comment (z-a-f) [reply torch/ao/nn/sparse/quantized/utils.py:22]: > > But I am still confused about why it's blocking landing of the current refactoring PR.
> 
> +1, I was only responding because Raghu tagged me, but this discussion is not related to the goal of this PR.

Oh I didn't mean to sound annoyed -- I welcome your contribution (and request it for other PRs as well). I just got surprised that this blew bigger than I thought it would :)

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465900).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465900).

File: test/ao/sparsity/test_kernels.py [hunk 2]
```diff
@@ -10,10 +10,9 @@
 import torch
 import torch.quantization as tq
 
-# Hopefully, the `mo` namespace will move under `nn`
 from torch import nn
 from torch.ao.nn.sparse import quantized as ao_nn_sq
-from torch.ao.nn.sparse.quantized.utils import QNNPACKLinearBlockSparsePattern
+from torch.ao.nn.sparse.quantized.utils import LinearBlockSparsePattern
```
Comment (raghuramank100) [inline test/ao/sparsity/test_kernels.py:15]: Please add a TODO: In a later PR where we modify from_float to take the sparsity pattern as an argument and this pattern will not be needed.

**Review by raghuramank100 — APPROVED** [review]

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465900).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465900).
